File: app/src/main/java/com/eam/investire/activity/ReceitasActivity.java
package com.eam.investire.activity;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.eam.investire.R;
import com.eam.investire.config.ConfiguracaoFirebase;
import com.eam.investire.helper.Base64Custom;
import com.eam.investire.helper.DateCustom;
import com.eam.investire.model.Movimentacao;
import com.eam.investire.model.Usuario;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

import java.util.Objects;

public class ReceitasActivity extends AppCompatActivity {

    private TextInputEditText campoData, campoCategoria, campoDescricao;
    private EditText campoValor;
    private final DatabaseReference firebaseRef = ConfiguracaoFirebase.getFirebaseDatabase();
    private final FirebaseAuth autenticacao = ConfiguracaoFirebase.getFirebaseAutenticacao();
    private double receitaTotal;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_receitas);

        campoData = findViewById(R.id.editData);
        campoCategoria = findViewById(R.id.editCategoria);
        campoDescricao = findViewById(R.id.editDescricao);
        campoValor = findViewById(R.id.editValor);

        campoData.setText(DateCustom.dataAtual());
        recuperarReceitaTotal();
    }

    public void salvarReceita(View view) {
        if (validarCamposReceita()) {
            Movimentacao movimentacao = new Movimentacao();
            String data = Objects.requireNonNull(campoData.getText()).toString();
            double valorRecuperado = Double.parseDouble(campoValor.getText().toString());

            movimentacao.setValor(valorRecuperado);
            movimentacao.setCategoria(Objects.requireNonNull(campoCategoria.getText()).toString());
            movimentacao.setDescricao(Objects.requireNonNull(campoDescricao.getText()).toString());
            movimentacao.setData();
            movimentacao.setTipo("r");

            double receitaAtualizada = receitaTotal + valorRecuperado;

            atualizarReceita(receitaAtualizada);
            movimentacao.salvar(data);
            finish();
        }
    }

    public Boolean validarCamposReceita() {
        String textoValor = campoValor.getText().toString();
        String textoData = Objects.requireNonNull(campoData.getText()).toString();
        String textoCategoria = Objects.requireNonNull(campoCategoria.getText()).toString();
        String textoDescricao = Objects.requireNonNull(campoDescricao.getText()).toString();

        if (!textoValor.isEmpty()) {
            if (!textoData.isEmpty()) {
                if (!textoCategoria.isEmpty()) {
                    if (!textoDescricao.isEmpty()) {
                        return true;
                    } else {
                        Toast.makeText(ReceitasActivity.this,
                                "Descrição não foi preenchida!",
                                Toast.LENGTH_SHORT).show();
                        return false;
                    }
                } else {
                    Toast.makeText(ReceitasActivity.this,
                            "Categoria não foi preenchida!",
                            Toast.LENGTH_SHORT).show();
                    return false;
                }
            } else {
                Toast.makeText(ReceitasActivity.this,
                        "Data não foi preenchida!",
                        Toast.LENGTH_SHORT).show();
                return false;
            }
        } else {
            Toast.makeText(ReceitasActivity.this,
                    "Valor não foi preenchido!",
                    Toast.LENGTH_SHORT).show();
            return false;
        }
    }

    public void recuperarReceitaTotal() {
        String emailUsuario = Objects.requireNonNull(autenticacao.getCurrentUser()).getEmail();
        assert emailUsuario != null;
        String idUsuario = Base64Custom.codificarBase64(emailUsuario);
        DatabaseReference usuarioRef = firebaseRef.child("usuarios").child(idUsuario);

        usuarioRef.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Usuario usuario = dataSnapshot.getValue(Usuario.class);
                assert usuario != null;
                receitaTotal = usuario.getReceitaTotal();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    public void atualizarReceita(double receita) {
        String emailUsuario = Objects.requireNonNull(autenticacao.getCurrentUser()).getEmail();
        assert emailUsuario != null;
        String idUsuario = Base64Custom.codificarBase64(emailUsuario);
        DatabaseReference usuarioRef = firebaseRef.child("usuarios").child(idUsuario);

        usuarioRef.child("receitaTotal").setValue(receita);
    }

    public void salvarDespesa(View view) {
    }
}

File: app/src/main/java/com/eam/investire/activity/ReceitasActivity.kt
package com.eam.investire.activity

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.eam.investire.R
import com.eam.investire.config.ConfiguracaoFirebase
import com.eam.investire.helper.Base64Custom
import com.eam.investire.helper.DateCustom
import com.eam.investire.model.Movimentacao
import com.eam.investire.model.Usuario
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import java.util.*

class ReceitasActivity : AppCompatActivity() {
    private var campoData: TextInputEditText? = null
    private var campoCategoria: TextInputEditText? = null
    private var campoDescricao: TextInputEditText? = null
    private var campoValor: EditText? = null
    private val firebaseRef = ConfiguracaoFirebase.firebaseDatabase
    private val autenticacao = ConfiguracaoFirebase.firebaseAutenticacao
    private var receitaTotal = 0.0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_receitas)
        campoData = findViewById(R.id.editData)
        campoCategoria = findViewById(R.id.editCategoria)
        campoDescricao = findViewById(R.id.editDescricao)
        campoValor = findViewById(R.id.editValor)
        campoData!!.setText(DateCustom.dataAtual())
        recuperarReceitaTotal()
    }

    fun salvarReceita(view: View?) {
        if (validarCamposReceita()) {
            val movimentacao = Movimentacao()
            val data = Objects.requireNonNull(campoData!!.text).toString()
            val valorRecuperado = campoValor!!.text.toString().toDouble()
            movimentacao.valor = valorRecuperado
            movimentacao.categoria = Objects.requireNonNull(campoCategoria!!.text).toString()
            movimentacao.descricao = Objects.requireNonNull(campoDescricao!!.text).toString()
            movimentacao.setData()
            movimentacao.tipo = "r"
            val receitaAtualizada = receitaTotal + valorRecuperado
            atualizarReceita(receitaAtualizada)
            movimentacao.salvar(data)
            finish()
        }
    }

    private fun validarCamposReceita(): Boolean {
        val textoValor = campoValor!!.text.toString()
        val textoData = Objects.requireNonNull(campoData!!.text).toString()
        val textoCategoria = Objects.requireNonNull(campoCategoria!!.text).toString()
        val textoDescricao = Objects.requireNonNull(campoDescricao!!.text).toString()
        return if (textoValor.isNotEmpty()) {
            if (textoData.isNotEmpty()) {
                if (textoCategoria.isNotEmpty()) {
                    if (textoDescricao.isNotEmpty()) {
                        true
                    } else {
                        Toast.makeText(this@ReceitasActivity,
                                "Descrição não foi preenchida!",
                                Toast.LENGTH_SHORT).show()
                        false
                    }
                } else {
                    Toast.makeText(this@ReceitasActivity,
                            "Categoria não foi preenchida!",
                            Toast.LENGTH_SHORT).show()
                    false
                }
            } else {
                Toast.makeText(this@ReceitasActivity,
                        "Data não foi preenchida!",
                        Toast.LENGTH_SHORT).show()
                false
            }
        } else {
            Toast.makeText(this@ReceitasActivity,
                    "Valor não foi preenchido!",
                    Toast.LENGTH_SHORT).show()
            false
        }
    }

    private fun recuperarReceitaTotal() {
        val emailUsuario = Objects.requireNonNull(autenticacao!!.currentUser)?.email!!
        val idUsuario = Base64Custom.codificarBase64(emailUsuario)
        val usuarioRef = firebaseRef!!.child("usuarios").child(idUsuario)
        usuarioRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val usuario = dataSnapshot.getValue(Usuario::class.java)!!
                receitaTotal = usuario.receitaTotal
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }

    private fun atualizarReceita(receita: Double) {
        val emailUsuario = Objects.requireNonNull(autenticacao!!.currentUser)?.email!!
        val idUsuario = Base64Custom.codificarBase64(emailUsuario)
        val usuarioRef = firebaseRef!!.child("usuarios").child(idUsuario)
        usuarioRef.child("receitaTotal").setValue(receita)
    }
}